src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp

```diff
@@ -104,321 +150,710 @@ enum Spells
+        void UpdateAI(uint32 diff) override
         {
-            if (bDone && uiResetTimer <= uiDiff)
-            {
-                me->GetMotionMaster()->MovePoint(0, 746.87f, 665.87f, 411.75f);
-                bDone = false;
-            } else uiResetTimer -= uiDiff;
+            events.Update(diff);
 
-            if (!UpdateVictim())
+            if (me->HasUnitState(UNIT_STATE_CASTING))
                 return;
 
-            if (uiHammerJusticeTimer <= uiDiff)
+            while (uint32 eventId = events.ExecuteEvent())
             {
-                me->InterruptNonMeleeSpells(true);
-
-                if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 250, true))
+                switch (eventId)
                 {
-                    if (target->IsAlive())
+                case EVENT_RADIANCE:
+                    DoCastAOE(SPELL_RADIANCE);
+                    Talk(EMOTE_RADIANCE);
+                    events.ScheduleEvent(EVENT_HAMMER_OF_JUSTICE, urand(20 * IN_MILLISECONDS, 30 * IN_MILLISECONDS));
+                    break;
+                case EVENT_HAMMER_OF_JUSTICE:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 100.0f, true))
                     {
-                        DoCast(target, SPELL_HAMMER_JUSTICE);
+                        DoCast(target, SPELL_HAMMER_JUSTICE, true);
                         DoCast(target, SPELL_HAMMER_RIGHTEOUS);
+                        Talk(EMOTE_HAMMER_RIGHTEOUS, target);
+                        Talk(SAY_HAMMER_RIGHTEOUS);
                     }
+                    events.ScheduleEvent(EVENT_RADIANCE, urand(7 * IN_MILLISECONDS, 15 * IN_MILLISECONDS));
+                    break;
+                case EVENT_CHALLENGE_DONE:
+                    me->RemoveAllAuras();
+                    me->GetThreatManager().ClearAllThreat();
+                    me->SetFullHealth();
+                    me->RestoreFaction();
+                    DoCastAOE(SPELL_EADRIC_ACH, true);
+                    DoCastAOE(SPELL_EADRIC_FACEROLLER, true);
+                    instance->DoUpdateAchievementCriteria(ACHIEVEMENT_CRITERIA_TYPE_BE_SPELL_TARGET, SPELL_EADRIC_ACH, 0, me);
+                    _JustDied();
+                    me->SetWalk(true);
+                    me->GetMotionMaster()->MovePoint(POINT_DESPAWN, bossExitPos);
+                    break;
+                default:
+                    break;
                 }
-                uiHammerJusticeTimer = 25000;
-            } else uiHammerJusticeTimer -= uiDiff;
-
-            if (uiVenganceTimer <= uiDiff)
-            {
-                DoCast(me, SPELL_VENGEANCE);
-
-                uiVenganceTimer = 10000;
-            } else uiVenganceTimer -= uiDiff;
-
-            if (uiRadianceTimer <= uiDiff)
-            {
-                DoCastAOE(SPELL_RADIANCE);
+            }
 
-                uiRadianceTimer = 16000;
-            } else uiRadianceTimer -= uiDiff;
+            if (!UpdateVictim())
```

Comment:
I really dislike `UpdateVictim` needing to be after the event handling loop. Can we change that somehow?